### ISSUE TITLE
Run fix for GUIDs, IDIRs & schedules missing on submissions to BI API services and ODS

### ISSUE THREAD

**Issue by Abuchana** [bcgov/digital-journeys#799]
- PDP Devs were able to recover both GUID, IDIR, and Schedule from KeyCloak and provided TES BI with a data file
- TES BI successfully uploaded the data file to the data warehouse in production
- The self-service report on HR Analytics that SHR uses now reflects the updated/corrected data
- Records for forms that were created prior to the fixes being implemented will still come in with null values
- We should expect to see entries with null values trickle in until all forms created prior to the fixes being implemented have either been submitted or deleted
- This process will need to be repeated to iteratively correct these records after they trickle in
- There are already a few records in the DWH with null values
- To repeat this process, someone on the PDP team will need to generate updated data files and provide those to TES BI
- This will need to be repeated until all draft forms created prior to the fixes are submitted or deleted - 60 days

**Comment by MeghanStothers** [comment]
To test after 60 days (since Nov 8 problem began) - We will fit in this work in January as we see fit. Not critical

**Comment by T854236** [comment]
@iman-jamali-fw 

Do we want to re-run this again this week for the missing GUID/IDIR and schedule information?

**Comment by iman-jamali-fw** [comment]
@T854236 

Sure, I can try later this week if that sounds good to you.

**Comment by Abuchana** [comment]
There are currently 1400 new Telework Agreements that have null values reporting to the data warehouse that need to be corrected. Please run fix again this week when time allows and does not conflict with prep work for 1.10 user testing that's planned.

**Comment by T854236** [comment]
@Abuchana @iman-jamali-fw 

New file loaded here: https://bcgov.sharepoint.com/:x:/r/teams/03991/Shared%20Documents/Data%20Collection/[Telework_submissions_with_missing_GUID_Jan6_2023.xlsx](https://bcgov.sharepoint.com/:x:/r/teams/03991/Shared%20Documents/Data%20Collection/Telework_submissions_with_missing_GUID_Jan6_2023.xlsx?d=wb72d91f563394fc7a3435f27fe52185e&csf=1&web=1&e=0ghtIp)?d=wb72d91f563394fc7a3435f27fe52185e&csf=1&web=1&e=0ghtIp

**Comment by Abuchana** [comment]
@T854236, thank you!!!

**Comment by iman-jamali-fw** [comment]
@T854236 

I send you an email containing the missing IDIR, GUID, and schedule data of the users you shared with me earlier. Hope this helps to fill the missing data :)

**Comment by iman-jamali-fw** [comment]
@T854236 

I just learned (while wokring on [#847](https://app.zenhub.com/workspaces/digital-journeys-61c10fe84c8bcb001491c5f5/issues/gh/bcgov/digital-journeys/847)) that some of the IDIRs I added to the that excel file (shared on sharepoint), don't have `@idir` at the end. 

Context: Some of the users's `idir` attributes we get when a user login with IDIR do not have `@idir` postfix, which is why Simen added a function to Keycloak to add that in case a user's `IDIR` don't have `@idir`. Since I extracted the `IDIR`s from Keycloak database, some of them don't have `@idir`. 

Is it going to be an issue on ODS side? I so, we need to add that postfix to the users without `@idir`.

**Comment by T854236** [comment]
@iman-jamali-fw 

It's ok, I can run a script to add if missing.

**Comment by T854236** [comment]
@iman-jamali-fw 

Yes thanks, I can make the change.  As an fyi, we also do URL encoding on the file.

Can you confirm what ID you are using to match lookup in the app?  The Excel spreadsheet has a column called application id that is higher than 25k, the file that you send with the schedule information is with appication_id >13k.  There is database table id that lines up to it and I wanted to confirm if the information you provided really is application_id or app_telework_info_id which is from ODS.

The reason I ask as when I merge data, it is joining value with some schedule that I don't expect to see.  For example, if the records had the NoFormalSchedule with a value of 1, I don't expect to see a schedule.

**Comment by iman-jamali-fw** [comment]
@T854236 

Oh no! I was using `app_telework_info_history_id` as the applicationId 😢

**Comment by iman-jamali-fw** [comment]
@T854236 

I resent the missing values with the correct `applicationId`s (in an email). Could you verify if those data are correct and can be saved to ODS? Thank you

**Comment by T854236** [comment]
@iman-jamali-fw 

Data was loaded successfully into PROD on January 19th in the evening.  We don't see any other records with missing gui/idir by applicationid in ODS.

**Comment by Abuchana** [comment]
@iman-jamali-fw and @T854236, Sarah is reporting more null values for the Telework Agreement. Can you have a look and let us know how we can remedy this? 
Thanks! Alison

I still see:
•	16 records with uncorrected NULL schedule when NoFormalSchedule is 0
•	One record with uncorrected null GUID/IDIR 

For reference, the queries I use to test this are below: 

--ApplicationID record with Max DWCreated has null schedule (where NoFormalSchedule is also 0)
SELECT [app_telework_info_history_id]
      ,[guid]
      ,[idir]
      ,[remote_date]
      ,[ThreeOrMoreDays]
      ,[admOrDelegateNameOpt]
      ,[fileNamesCSV]
      ,[NoFormalSchedule]
      ,[agreementBetweenMinistry]
      ,[manager_guid]
      ,[manager_idir]
      ,[managersSignatureTimestamp]
      ,[employeesCommentsOptional]
     ,[duties]
      ,[noteOptional]
      ,[action]
      ,[schedule]
      ,[signed_date]
      ,[applicationStatus]
      ,[address1]
      ,[city]
      ,[postal]
      ,[numberOfDaysTeleworkingInAWeek]
      ,[employeeSignatureTimestamp]
      ,[read]
      ,[acknowledged]
      ,[applicationId]
      ,[date_created]
      ,[DwEMPLID]
      ,[DwEMPL_RCD]
      ,[DwDEPTID]
      ,[DwManagerEmplID]
      ,[DwDaysInOffice]
      ,[DWCreated]
FROM [Datamart_TeleWork].[Telework].[app_telework_info_history] AS A
WHERE 
       DWCreated = (SELECT MAX (B.DWCreated) FROM [Datamart_TeleWork].[Telework].[app_telework_info_history] AS B
              WHERE A.applicationId = B.applicationId)
                       AND [schedule] IS NULL
                       AND [NoFormalSchedule] = 0;

--ApplicationID record with Max DWCreated has NULL IDIR or GUID
SELECT [app_telework_info_history_id]
      ,[guid]
      ,[idir]
      ,[remote_date]
      ,[ThreeOrMoreDays]
      ,[admOrDelegateNameOpt]
      ,[fileNamesCSV]
      ,[NoFormalSchedule]
      ,[agreementBetweenMinistry]
      ,[manager_guid]
      ,[manager_idir]
      ,[managersSignatureTimestamp]
      ,[employeesCommentsOptional]
      ,[duties]
      ,[noteOptional]
      ,[action]
      ,[schedule]
      ,[signed_date]
      ,[applicationStatus]
      ,[address1]
      ,[city]
      ,[postal]
      ,[numberOfDaysTeleworkingInAWeek]
      ,[employeeSignatureTimestamp]
      ,[read]
      ,[acknowledged]
      ,[applicationId]
      ,[date_created]
      ,[DwEMPLID]
      ,[DwEMPL_RCD]
      ,[DwDEPTID]
      ,[DwManagerEmplID]
      ,[DwDaysInOffice]
     ,[DWCreated]
FROM [Datamart_TeleWork].[Telework].[app_telework_info_history] AS A
WHERE 
       DWCreated = (SELECT MAX (B.DWCreated) FROM [Datamart_TeleWork].[Telework].[app_telework_info_history] AS B
              WHERE A.applicationId = B.applicationId)
                       AND ([IDIR] IS NULL OR [GUID] IS NULL);

Thanks,
Sarah

Sarah Miles, Workforce Intelligence Analyst

**Comment by iman-jamali-fw** [comment]
@T854236 

Could you provide the list of these affected submissions? :)

**Comment by T854236** [comment]
@iman-jamali-fw 

Loaded a new file, 17 records total.

https://bcgov.sharepoint.com/:x:/r/teams/03991/Shared%20Documents/Data%20Collection/[Telework_submissions_with_missing_schedule_or_guid_Feb8_2023.csv](https://bcgov.sharepoint.com/:x:/r/teams/03991/Shared%20Documents/Data%20Collection/Telework_submissions_with_missing_schedule_or_guid_Feb8_2023.csv?d=w021856bc0bb640fa94b7fc39edf01ed7&csf=1&web=1&e=kiaRNG)?d=w021856bc0bb640fa94b7fc39edf01ed7&csf=1&web=1&e=kiaRNG

**Comment by Stella-Archer** [comment]
@bhumin-fw  could you take a look at this and comment on the necessity of doing this work?
